### PR TITLE
fix: prevent _sync_dist deadlock for dist_reduce_fx=None list states

### PR DESCRIPTION
**Pull request by jlaportebot**
## Problem

When a metric uses `dist_reduce_fx=None` with a list state that is empty on some ranks, `_sync_dist` would try to gather zero-length tensors, causing a distributed deadlock because `torch.distributed` requires all ranks to participate with matching tensor sizes.

Fixes #3336

## Solution

- Skip the gather entirely for `None`-reduce states (they are not reduced across ranks anyway)
- Filter out empty placeholder tensors before stacking
- Correct indentation in the post-gather loop that was misaligned

## Note

Clean rebase of the previously closed #3390, which had a bloated pre-commit.ci commit.

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/metric.py [hunk 1]
```diff
@@ -506,14 +506,23 @@ def _sync_dist(self, dist_sync_fn: Callable = gather_all_tensors, process_group:
...
-            # cornor case in distributed settings where a rank have not received any data, create empty to concatenate
-            if (
-                self._TORCH_GREATER_EQUAL_2_1
-                and reduction_fn == dim_zero_cat
-                and isinstance(input_dict[attr], list)
-                and len(input_dict[attr]) == 0
-            ):
-                input_dict[attr] = [torch.tensor([], device=self.device, dtype=self.dtype)]
+        # corner case in distributed settings where a rank have not received any data, create empty to concatenate
+        if (
```
Comment (justusschock) [inline src/torchmetrics/metric.py]: This needs to be within the loop body
